test_runner/src/main/kotlin/ftl/http/ExecuteWithRetry.kt

```diff
@@ -2,44 +2,60 @@ package ftl.http
 
 import com.google.api.client.googleapis.services.json.AbstractGoogleJsonClientRequest
 import com.google.api.client.http.HttpResponseException
+import com.google.api.client.http.HttpStatusCodes
 import ftl.config.FtlConstants
 import ftl.util.PermissionDenied
 import ftl.util.ProjectNotFound
 import kotlinx.coroutines.delay
 import kotlinx.coroutines.runBlocking
 import java.io.IOException
 import kotlin.math.exp
-import kotlin.math.roundToInt
+import kotlin.math.roundToLong
 
 // Only use on calls that don't change state.
 // Fetching status is safe to retry on timeout. Creating a matrix is not.
 fun <T> AbstractGoogleJsonClientRequest<T>.executeWithRetry(): T = withRetry { this.execute() }
 
 private inline fun <T> withRetry(crossinline block: () -> T): T = runBlocking {
-    var lastErr: IOException? = null
-
-    repeat(4) {
-        try {
-            return@runBlocking block()
-        } catch (err: IOException) {
-            // We want to send every occurrence of Google API error for statistic purposes
-            // https://github.com/Flank/flank/issues/701
-            FtlConstants.bugsnag?.notify(FlankGoogleApiError(err))
-
-            lastErr = err
-            if (err is HttpResponseException) {
-                // we want to handle some FTL errors with special care
-                when (err.statusCode) {
-                    429 -> return@repeat
-                    403 -> throw PermissionDenied(err)
-                    404 -> throw ProjectNotFound(err)
-                }
-            }
-            delay(exp(it - 1.0).roundToInt().toLong())
-        }
+    var lastError: IOException? = null
+    repeat(maxTries) { repeatCounter ->
+        val executionResult = tryExecuteBlock(block)
+        if (executionResult.isSuccess()) return@runBlocking executionResult.success()
+
+        val error = executionResult.error()
+        // We want to send every occurrence of Google API error for statistic purposes
+        // https://github.com/Flank/flank/issues/701
+        FtlConstants.bugsnag?.notify(FlankGoogleApiError(error))
+        lastError = error
+        if (handleFtlError(error)) return@repeat
+        delay(calculateDelayTime(repeatCounter))
     }
+    throw IOException("Request failed", lastError)
+}
+private const val maxTries = 4
+
+private inline fun <T> tryExecuteBlock(block: () -> T): ExecutionBlockResult<T> = try {
+    ExecutionBlockResult(block(), null)
+} catch (ioError: IOException) {
+    ExecutionBlockResult(null, ioError)
+}
 
-    throw IOException("Request failed", lastErr)
+private data class ExecutionBlockResult<T>(private val success: T?, private val error: IOException?) {
+    fun isSuccess() = success != null
+    fun success() = success!!
+    fun error() = error!!
 }
```

Comment:
This commit brings 16 lines of unnecessary abstraction. Code does nothing more then wrapping `try catch`, but is much harder to understand. Pls revert it. The only one reasonable change which I see here is removing redundant`return@runBlocking` from `return@runBlocking block()`.
BTW kotlin already have functional abstraction for `try catch` it's `runCatching`, but it doesn't bring much more to error handling IMO is usefully only for specific cases.

Reply:
I think @jan-gogo is right, we should revert this change and maybe think about different approach

Reply:
Thanks for feedback! I revert changes